### PR TITLE
Dice coefficient: sort pairs by similarity

### PR DESCRIPTION
**Pull request by nbgl**
Naive solution to Dice coefficient code not sorting candidate pairs, as described in #135. This PR improves accuracy but decreases speed—both these effects are significant in some cases but negligible in others.

~~Please do not merge this PR as the issue is still under discussion.~~

Edit: consensus is to make this change. The changes made in this PR can be optimised later if need be.

### CODE REVIEW

File: anonlink/similarities/_dice_python.py [hunk 1]
```diff
@@ -64,5 +65,21 @@ def dice_coefficient_python(
         result_sims.extend(sim for _, sim in top_k)
         result_indices0.extend(repeat(i, len(top_k)))
         result_indices1.extend(j for j, _ in top_k)
+
+    np_sims = np.frombuffer(result_sims, dtype=result_sims.typecode)
```
Comment (hardbyte) [inline anonlink/similarities/_dice_python.py]: If we have to use numpy - why not use numpy arrays from the start?

File: anonlink/similarities/_dice_x86.py [hunk 2]
```diff
@@ -111,4 +112,20 @@ def dice_coefficient_accelerated(
...
+    
+    np.negative(np_sims, out=np_sims)  # Sort in reverse.
+    # Mergesort is stable. We need that for correct tiebreaking.
+    order = np.argsort(np_sims, kind='mergesort')
+    np.negative(np_sims, out=np_sims)
+
+    # This modifies the original arrays since they share a buffer.
+    np_sims[:] = np_sims[order]
+    np_indices0[:] = np_indices0[order]
+    np_indices1[:] = np_indices1[order]
```
Comment (hardbyte) [inline anonlink/similarities/_dice_x86.py]: This is a bit of dejavu from the previous code section. However I assume you want to leave it separate so we can more easily optimize.

File: anonlink/similarities/_dice_python.py (hunk 1 above)
Comment (nbgl) [reply anonlink/similarities/_dice_python.py]: NumPy arrays are not dynamic. This makes it harder to use them to store a variable number of candidates.

The problem with just casting a Python array to NumPy is that NumPy won't take ownership of the memory, so we'd have issues with garbage collection.

File: anonlink/similarities/_dice_x86.py (hunk 2 above)
Comment (nbgl) [reply anonlink/similarities/_dice_x86.py]: Hmmm, maybe it'd be better to put it in a separate function for now…

**Comment by nbgl** [comment]
@hardbyte could you please have a brief glance at the changes I've made since your review?

**Pull request merged**